src/tribler/core/components/reporter/exception_handler.py

```diff
@@ -134,6 +145,12 @@ def unhandled_error_observer(self, _, context):
             if process_manager:
                 process_manager.current_process.set_error(exception)
 
+            # If should_stop is True, the error is critical, so we first save it to the file to ensure
+            # that it will be reported to the GUI via events endpoint either immediately or later
+            # after core restart. This saved file will be deleted when the error is reported to the GUI.
+            if should_stop:
```

Comment:
From my understanding, if error reports are handled by the exception handler, then they will be sent to the GUI without any problem.

Could you please provide a way to reproduce the initial problem that this PR is aimed to solve?

Reply:
To reproduce a part of the solution, `CTRL` + `ALT` +  `SHIFT` + `C` will trigger a core error which is sent to the GUI as an error report and in turn restarts the core. The core_restart_logs should be visible on sentry then.

The other part where the core exception is logged and sent to sentry on the next restart is difficult to reproduce. To test this, I manually placed a saved exception on the crash_logs directory inside the state directory.

Reply:
> To reproduce a part of the solution, CTRL + ALT + SHIFT + C will trigger a core error which is sent to the GUI as an error report and in turn restarts the core. The core_restart_logs should be visible on sentry then.

Should I press it within the Tribler GUI? Is it `Ctrl+Opt+Shift+C` on macOS?

Reply:
It is a keyboard shortcut to trigger core exception.
https://github.com/Tribler/tribler/blob/8bb85b50d5d0c4a0580630820079974b98e5f2f9/src/tribler/gui/tribler_window.py#L237-L241

Reply:
Basically, this combination just calls `core_exception_handler.unhandled_error_observer`, which is not a way to reproduce the real use case of the improvements proposed in this PR.

```python
    async def core_test_exception(self, _):
        self._logger.info('Test Tribler Core exception was triggered')
        if self.core_exception_handler:
            try:
                raise TriblerCoreTestException('Tribler Core Test Exception')
            except TriblerCoreTestException as e:
                context = dict(should_stop=True, message='Test message', exception=e)
                self.core_exception_handler.unhandled_error_observer(None, context)
        else:
            self._logger.info('Exception handler is not set in DebugEndpoint')

        return RESTResponse({"success": True})
```

So for me, it is still not clear which case the improvement covers. I understand the theoretical point from the original issue but do not understand their relation with the implementation.

In practice, which type of errors could trigger the new logic and become visible for us but not visible with the old logic? Where should an error be raised in the core to prove the new logic adds value to error handling?

Reply:
> In practice, which type of errors could trigger the new logic and become visible for us but not visible with the old logic?

Here is the original issue this PR is developed to tackle:
- https://github.com/Tribler/tribler/issues/7603 

And, here is the linked sentry issue for it:
- https://sentry.tribler.org/organizations/tribler/issues/2003/events/f1a9c56138aa4dbea5598d00c2a72311/?project=7&query=is%3Aunresolved+version%3A7.13.0&referrer=previous-event&stream_index=2

In the case where the Core connection is lost, we do not know what caused the issue. From the uptime, we see that Tribler was running for 7205 seconds. Then suddenly the core connection was lost and we have no idea what caused it. The original error in the core that broke the GUI connection is unknown. 

With this PR, such a core error should be logged as a file (assuming the exception handler on the core catches such an exception) and sent to the GUI when the core connection is re-established after the core restart. Then the error report sent by the user should give us more information about the original cause of the previous exception.

> Where should an error be raised in the core to prove the new logic adds value to error handling?

Regarding the reproduction of such a case practically under a normal Tribler run, I'm not able to do so. Which is why the above suggestion was mentioned.

Reply:
I understand your intention and appreciate it. I'm just questioning the implementation, particularly whether the suggested change could reveal the real case of https://github.com/Tribler/tribler/issues/7603.

I'm uncertain about the change because we actually cannot imagine the potential case for the error; therefore, we cannot verify the solution, even theoretically. It is such a significant change you propose (15 new commits, +700 LOC) and it changes the interaction between Core and GUI pretty significantly, which could lead to unforeseen circumstances. I'm not saying the PR is good or bad; I'm just highlighting the risks and questioning the worthiness of taking them.

From my POV, if the error is successfully handled by the core's error handler, then it should be sent to the GUI without any problem, so the fix you suggested might be insufficient.

One of the explanations for https://github.com/Tribler/tribler/issues/7603 could be the possibility that someone (maybe even a user) just killed the process following the observation that the process consumes a lot of CPU and/or memory resources.

Another explanation could be the probability of some weird bug in `aiohttp` that led to an HTTP server crash, which in turn led to the `CoreConnectionError`. With this explanation, the suggested solution could work, but we need to verify it somehow.

Reply:
I came up with two more probable scenarios that could lead to a `CoreConnectionError` but are not related to `aiohttp`. I'll explain them tomorrow.

Reply:
I've added scenarios here: https://github.com/Tribler/tribler/issues/7647#issuecomment-1838527794